### PR TITLE
feat: implement GraphQL over WebSocket Protocol

### PR DESCRIPTION
**Pull request by amondnet**
Pull request checklist
----

- [x] Please read our [contributor guide](https://github.com/Netflix/dgs-framework/blob/master/CONTRIBUTING.md)
- [x] Consider creating a discussion on the [discussion forum](https://github.com/Netflix/dgs-framework/discussions)
  first
- [x] Make sure the PR doesn't introduce backward compatibility issues
- [ ] Make sure to have sufficient test cases

Pull Request type
----

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe):

Changes in this PR
----

_Describe the new behavior from this PR, and why it's needed_
Issue https://github.com/Netflix/dgs-framework/discussions/684

Alternatives considered
----

_Describe alternative implementation you have considered_

### CODE REVIEW

**Comment by srinivasankavitha** [comment]
What set up are you using for the client to test the server? Are you using Apollo client with `graphql-ws` library?

File: graphql-dgs-transports-websockets/src/main/kotlin/com/netflix/graphql/dgs/transports/websockets/DgsWebsocketTransport.kt [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+                        objectMapper.writeValueAsBytes(
+                            Message.PongMessage(
+                                payload = message.payload
+                            )
+                        )
+                    )
+                )
+            }
+            is Message.PongMessage -> {
+                // TODO: onPong
```
[the remark this answers is not in the document]
Comment (amondnet) [reply graphql-dgs-transports-websockets/src/main/kotlin/com/netflix/graphql/dgs/transports/websockets/DgsWebsocketTransport.kt:84]: @srinivasankavitha According to the spec, it should call the onPong event handler if exists.

**Comment by srinivasankavitha** [comment]
@amondnet - thanks for adding the test and that looks good. The PR overall looks good to me.  I tried  to test this with a real client with the DGS based on the docs here: https://github.com/enisdenjo/graphql-ws#use-the-client but it does not seem to work after enabling the new module. Have you been able to set up a client to test this end to end? Could you share what you have, if so? Thanks!

**Comment by michaelboyles** [comment]
Happy to test my (closed source) application with an Apollo client if you like. Is there any config necessary on the DGS side, or does the client negotiate the subprotocol as part of the handshake or something?

**Comment by srinivasankavitha** [comment]
https://github.com/Netflix/dgs-framework/pull/686#issuecomment-1066212140
- That would be really helpful! Unfortunately, the testing will involve cloning and building the forked repo with the changes and publishing a local snapshot that you can use in your DGS app. You will also need to add the new module `implementation(project(":graphql-dgs-transports-websockets"))` in your DGS app to get this functionality.

**Comment by srinivasankavitha** [comment]
> Happy to test my (closed source) application with an Apollo client if you like. Is there any config necessary on the DGS side, or does the client negotiate the subprotocol as part of the handshake or something?

That would be really helpful! Unfortunately, the testing will involve cloning and building the forked repo with the changes and publishing a local snapshot that you can use in your DGS app. You will also need to add the new module implementation(project(":graphql-dgs-transports-websockets")) in your DGS app to get this functionality.

**Comment by amondnet** [comment]
@paulbakker 



> Can this be used together with the existing implementation? E.g. can both modules be active at the same time? It does look like it, because the new module is registering on /graphql.

Yes, we can use together. `Websocket` uses `http` only during handshake.
https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#the_websocket_handshake

> Both WebMVC and Webflux should be supported

Ok. I will add both.

> Does this conflict with "normal" queries since it's on /graphql? (I think this should work, but we need to make sure).

No conflict.

File: settings.gradle.kts [hunk 2]
```diff
@@ -39,3 +39,5 @@ include("graphql-dgs-webflux-starter")
 include("graphql-dgs-example-shared")
 include("graphql-dgs-pagination")
 include("graphql-dgs-subscription-types")
+include("graphql-dgs-transports-websockets")
+include("graphql-dgs-transports-websockets-autoconfigure")
```
[the remark this answers is not in the document]
Comment (amondnet) [reply settings.gradle.kts]: @paulbakker It can be used with webmvc and webflux.

**Comment by michaelboyles** [comment]
Sorry for the delay. I'm still happy to test this if it will help push it through. Unfortunately it's not currently building because of linter errors.

**Comment by srinivasankavitha** [comment]
@michaelboyles - I am now able to test this with Apollo client and have some fixes. Will push those in the next day. I will be doing some more testing and likely get this PR merged after next week. Will post updates here. Apologies again for the delay due to conflicting internal priorities.

**Comment by srinivasankavitha** [comment]
@amondnet - could you please give me access to push a branch to your fork? I have some fixes based on testing your PR.

**Comment by amondnet** [comment]
@srinivasankavitha I have invited you as a collaborator.

**Comment by amondnet** [comment]
https://github.com/Netflix/dgs-framework/tree/feat-subscriptions-graphql-ws

**Comment by srinivasankavitha** [comment]
Merged this PR with additional fixes, implementation for webflux and some refactoring: https://github.com/Netflix/dgs-framework/pull/1200.

I'll close this one out. Thanks @amondnet for all the contributions for this feature.

**Pull request closed without merging**